app/concerns/sortable_events.rb

```diff
@@ -0,0 +1,176 @@
+module SortableEvents
+  extend ActiveSupport::Concern
+
+  included do
+    validate :validate_events_order
+  end
+
+  def description_events_order(events = 'events created in each run')
+    <<-MD.lstrip
+        To specify the order of #{events}, set `events_order` to an array of sort keys, each of which looks like either `expression` or `[expression, type, descending]`, as described as follows:
+
+        * _expression_ is a Liquid template to generate a string to be used as sort key.
+
+        * _type_ (optional) is one of `string` (default), `number` and `time`, which specifies how to evaluate _expression_ for comparison.
+
+        * _descending_ (optional) is a boolean value to determine if comparison should be done in descending (reverse) order, which defaults to `false`.
+
+        Sort keys listed eariler take precedence over ones listed later.  For example, if you want to sort articles by the date and then by the author, specify `[["{{date}}", "time"], "{{author}}"]`.
+
+        Sorting is done stably, so even if all events have the same set of sort key values the original order is retained.  Also, a special Liquid variable `_index_` is provided, which contains the zero-based index number of each event, which means you can exactly reverse the order of events by specifying `[["{{_index_}}", "number", true]]`.
+    MD
+  end
+
+  module ClassMethods
+    def can_order_created_events!
+      raise if cannot_create_events?
+      prepend AutomaticSorter
+    end
+
+    def can_order_created_events?
+      include? AutomaticSorter
+    end
+
+    def cannot_order_created_events?
+      !can_order_created_events?
+    end
+  end
+
+  def can_order_created_events?
+    self.class.__send__(__callee__)
+  end
+
+  def cannot_order_created_events?
+    self.class.__send__(__callee__)
+  end
+
+  def events_order
+    options['events_order']
+  end
+
+  module AutomaticSorter
+    def check
+      return super unless events_order
+      sorting_events do
+        super
+      end
+    end
+
+    def receive(incoming_events)
+      return super unless events_order
+      # incoming events should be processed sequentially
+      incoming_events.each do |event|
+        sorting_events do
+          super([event])
+        end
+      end
+    end
+
+    def create_event(attrs)
+      if @sortable_events
+        @sortable_events << events.build({ user: user }.merge(attrs))
+      else
+        super
+      end
+    end
+
+    private
+
+    def sorting_events(&block)
+      @sortable_events = []
+      yield
+    ensure
+      events, @sortable_events = @sortable_events, nil
+      sort_events(events).each do |event|
```

Comment:
I think `expires_at` should be set based on when the event is actually emitted instead of when create_event() is called (to be buffered for later reordering), so I'm setting it right before `event.save!`.